### ISSUE TITLE
Add pkgconfig file in the sources

### ISSUE THREAD

**Issue by sl1pkn07** [KhronosGroup/glslang#1715]
Hi

is possible add a pkgconfig file in the sources, like vulkan-loader and spirv-tools?

greetings

**Comment by sl1pkn07** [comment]
Hi

any notice about this?

greetings

**Comment by smcv** [comment]
Generating pkg-config metadata would be a useful way to codify what is necessary to link to each of the API entry points provided by glslang. Sections of the API that are identifiably different (like the C++ class interface vs. the C functional interface, or any features that pull in additional dependencies) might benefit from separate .pc files.

At the moment, users of glslang need to know the names of all the static libraries, as well as the dependencies between them, and the fact that some (all?) need `-lpthread` on platforms that have it. pkg-config metadata could reduce this to something like:

```
# CMake
pkg_check_modules (GLSLANG REQUIRED glslang>=7.10)
# ... compile with ${GLSLANG_INCLUDEDIR}, link with ${GLSLANG_LIBRARIES}

# or Meson
glslang = dependency('glslang', version : '>=7.10')
executable(..., dependencies : [glslang])

# or Autotools
PKG_CHECK_MODULES([GLSLANG], [glslang >= 7.10])
# ... compile with ${GLSLANG_CFLAGS}, link with ${GLSLANG_LIBS}

# or plain Make/shell
c++ -oexample example.cpp $(pkg-config --cflags --libs "glslang >= 7.10")
```

**Comment by smcv** [comment]
The Fedora packaging at https://src.fedoraproject.org/rpms/glslang/tree/master seems to have patches for this, which have also been applied in Debian.

**Comment by DanySpin97** [comment]
Pull request #2083 upstream Fedora patch.

**Comment by jpr42** [comment]
I mostly* got this working. Except for Windows. See: https://github.com/KhronosGroup/glslang/pull/3898#issuecomment-2709189001

**Comment by jpr42** [comment]
glslang should just adopt CPS files instead:
https://www.kitware.com/common-package-specification-is-out-the-gate/

Supporting *.pc files with CMake is never going to be well supported and you are just going to get more issues.

CPS is the future standard.

**Comment by jpr42** [comment]
Ultimately supporting CPS will support other build systems other than CMake which is the true point of this issue.

Supporting pkgconfig will only cause problems at this point. Especially since it won't work well on all platforms/situations.

**Comment by sertonix** [comment]
Meaning it would not work well with most build systems except cmake? At least godot is not using cmake, but calling pkg-config would work well for it.

**Comment by jpr42** [comment]
> Meaning it would not work well with most build systems except cmake? At least godot is not using cmake, but calling pkg-config would work well for it.

CPS is intended to work with any build system. Not just CMake.

**Comment by sertonix** [comment]
Why would other build systems implement CPS when they already have good .pc support? And which of the .pc issues are so non-fixable that it needs a new standard?

**Comment by jpr42** [comment]
> And which of the .pc issues are so non-fixable that it needs a new standard?

See my PR for my attempt at trying to get CMake to generate it properly.

CPS team has done a write up on why .pc files are problematic:
https://cps-org.github.io/cps/overview.html
https://cps-org.github.io/cps/history.html <- Relevant details on pkg-config (I'm sure there are long discussions about this on the GitHub if you are truly interested)

>Why would other build systems implement CPS when they already have good .pc support?

It's not just CMake that will be adopting the standard.

https://github.com/cps-org/cps

**Comment by jpr42** [comment]
My understanding IIRC is that the following tools (CMake, Conan, Meson, Autotools, and vcpkg) will/currently support CPS. With more to come.

**Comment by kaniini** [comment]
A standard which is not yet actually adopted outside CMake cannot be described as 'the future' regardless of how aggressively Kitware markets it.

Autotools and Meson continue to use pkg-config.

**Comment by kaniini** [comment]
What I will say is that there is some limited amount of interest in supporting CPS metadata in pkgconf, but this "us vs. them" mentality ain't it.

**Comment by jpr42** [comment]
Understood.

I'm largely coming from the angle that adding pkg-config support won't work well. 

See my original PR: https://github.com/KhronosGroup/glslang/pull/3898

I realized it had several unsolvable problems: https://github.com/KhronosGroup/glslang/pull/3898#issuecomment-2709189001

**Comment by kaniini** [comment]
Yes, I agree, the pkg-config ecosystem is hard to make use of in CMake.  The pkgconf project still has some grant money left over from the SBOM work, I might see if anyone is interested in improving the situation from our side.  Maybe we can write a macro package that makes life better for CMake users.

And, like I said, there is interest from our side in at least supporting CPS metadata as an alternative metadata format.  And indeed, semantic pc(5) is shaped much more like CPS than the original format.  But it will take a few years for CPS or semantic pc(5) to really gain adoption in the real world, as the tooling needs to mature for both formats.

We should be focusing on getting developers access to the metadata they need, rather than debating a format war :)

**Comment by jpr42** [comment]
> Yes, I agree, the pkg-config ecosystem is hard to make use of in CMake.

Yes it's hard. I had various issues with maintaining it on several popular CMake projects.

> The pkgconf project still has some grant money left over from the SBOM work, I might see if anyone is interested in improving the situation from our side. Maybe we can write a macro package that makes life better for CMake users

Would be great if that got into CMake upstream. If CMake could easily generate pkg-config that would be lovely.

> We should be focusing on getting developers access to the metadata they need, rather than debating a format war :)

Agreed.

**Comment by kaniini** [comment]
It is unlikely that Kitware would add pkg-config support directly to CMake.  But Autotools also itself does not have direct pkg-config support: we ship a macro package (pkg.m4) which integrates the pkg-config ecosystem into Autotools.  We can do the same for CMake.

And, eventually, pkgconf community & Kitware will come to enough agreement on CPS that pkgconf adopts CPS as a supported metadata format.  But there is more work to be done there: CPS is largely CMake-focused, while the semantic pc(5) work we are doing tries to bring CPS-like features into pkg-config ecosystem to make authoring pc(5) files a lot easier.

tl;dr: applying "stop energy" to promote CPS is not helpful to anyone, including to CPS.  It just hurts users who are on Unix-like systems.

**Comment by jpr42** [comment]
I didn't mean to come off as anti pkg-config. Just trying to keep maintaining CMake code easy/correct.

I have no preference.

**Comment by kaniini** [comment]
Yep, +1 on that.  We should make an official pkg-config macro package like we have with Autotools.  That is easy to do.
